src/clients/python/experimental_api_v2/library/grpcclient.py

```diff
@@ -1174,3 +1239,188 @@ def get_response(self, as_json=False):
             return json.loads(MessageToJson(self._result))
         else:
             return self._result
+
+
+class InferStream:
+    """Supports sending inference requests and receiving corresponding
+    requests on a gRPC bi-directional stream.
+
+    Parameters
+    ----------
+    callback : function
+        Python function that is invoked upon receiving response from
+        the underlying stream. The function must have last argument
+        as **kwargs. The client may provide upto three keyworded
+        arguments in **kwargs. For the successful inference, InferResult
+        object will be provided with key 'result'. If there is an
+        error in inference, InferenceServerException object will be
+        provided with key 'error'. If this response is part of a
+        sequence then the sequence ID is provided with the key 
+        'sequence_id'. The 'sequence_id' may not be provided if
+        it was not resolved before an error occured.
+    """
+
+    def __init__(self, callback):
+        self._callback = callback
+        self._request_queue = queue.Queue()
+        self._handler = None
+        self._headers = None
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self.close()
+
+    def __del__(self):
+        self.close()
+
+    def close(self):
+        """Gracefully close underlying gRPC streams. Note that this call
+        blocks till response of all currently enqueued requests are not
+        received.
+        """
+        if self._is_initialized():
+            self._request_queue.put(None)
+            if self._handler.is_alive():
+                self._handler.join()
+            self._handler = None
+
+    def set_headers(self, headers):
```

Comment:
Is this really necessary? There is already a way to set headers for individual requests, why also on the stream?